### ISSUE TITLE
老哥，实验结果对不上啊

### ISSUE THREAD

**Issue by Bhongcheng** [JianXu95/FedPAC#2]
老哥，我按照你论文里的设置（如不同数据集有各自的学习率和batch size），运行后观察最后一轮的准确率，发现结果与论文对不上啊，我不知道哪里出现了问题。
这是论文的结果
![image](https://user-images.githubusercontent.com/90205287/236839939-59d40d23-efe3-44d0-b942-ba9657eba3b8.png)
这是我记录的部分结果
![image](https://user-images.githubusercontent.com/90205287/236840148-dc242bdc-63ac-47d9-9d10-d4e05f56a2e5.png)
前两个我改了seed，fmnist完全就是下载老哥你的代码后运行的，
emnist的实验我还在运行

为啥会这样呢呀。。。老哥你能否在百忙之中跑一下代码看看结果呢。。。。。。。

万分感激！

**Comment by JianXu95** [comment]
虽然我release的和实际最初用的不完全一样，也不可能差这么多吧，肯定哪儿有点问题😂fmnist 纯 local都不可能那么低，我有空时候看看发自我的 iPhone在 2023年5月8日，06:47，鲍鸿丞 ***@***.***> 写道：﻿
老哥，我按照你论文里的设置（如不同数据集有各自的学习率和batch size），运行后观察最后一轮的准确率，发现结果与论文对不上啊，我不知道哪里出现了问题。
这是论文的结果

这是我记录的部分结果

前两个我改了seed，fmnist完全就是下载老哥你的代码后运行的，
emnist的实验我还在运行
为啥会这样呢呀。。。老哥你能否在百忙之中也运行一下呢。。。。。。。
感谢老哥。。。。

**Comment by JianXu95** [comment]
我试了下fmnist没问题呀，差不多第38个round就能达到90%以上了


---- Global Communication Round : 38 ----
Train Loss: 0.23447757073057202, 0.2358179222792387
Local Accuracy on Local Data: 90.17500000000001%, 89.54166666666667%


| |
徐健
|
|
***@***.***
|
在2023年5月8日 ***@***.***> 写道：

老哥，我按照你论文里的设置（如不同数据集有各自的学习率和batch size），运行后观察最后一轮的准确率，发现结果与论文对不上啊，我不知道哪里出现了问题。
这是论文的结果

这是我记录的部分结果

前两个我改了seed，fmnist完全就是下载老哥你的代码后运行的，
emnist的实验我还在运行

为啥会这样呢呀。。。老哥你能否在百忙之中也运行一下呢。。。。。。。

感谢老哥。。。。

**Comment by Bhongcheng** [comment]
老哥老哥，我又重新下载然后只改命令行重跑结果，之前也不知道出了什么问题，这次的结果就正常多了
<img width="266" alt="image" src="https://user-images.githubusercontent.com/90205287/237022485-4e827606-40ca-46ff-b9e5-95e5a5d98041.png">
可是CINIC10的结果还是和第一次一样，和论文的结果对不上呀。。。

这是 CINIC-10数据集num为20下最后2轮的结果

---- Global Communication Round : 199 ----
(1) Agg Weights of Classifier Head
[0.925225783537992, 0.0026694361496219393, 0.0027665402158642068, 0.0, -0.0, 0.005284155492339917, -0.0, 0.0, 0.0, -0.0, 0.05888012258097735, -0.0, 0.0, 0.0, -0.0, 0.0, 0.0, -0.0, -0.0, 0.004165715733843089] 

Train Loss: 0.5214172199368478, 0.4143478119124969
Local Accuracy on Local Data: 63.80000000000001%, 58.7%

---- Global Communication Round : 200 ----
(1) Agg Weights of Classifier Head
[0.9213015966944889, 0.002600303230981748, 0.002391934643719771, 0.0, -0.0, 0.003465201882905373, 0.0, 0.0, 0.0010412572459559935, -0.0, 0.06475273348448668, -0.0, 0.0, -0.0, -0.0, 0.0, 0.0, 0.0, 0.0, 0.003975812471072751] 

Train Loss: 0.5193375659485658, 0.4137377354005972
Local Accuracy on Local Data: 63.5%, 59.166666666666664%

这个结果正常吗，而且分类器聚合权重全集中到第一个上了。。。

分隔线——————————————————————————————————————————————————————————

然后我还想麻烦老哥想请教一些问题：
<img width="657" alt="image" src="https://user-images.githubusercontent.com/90205287/237023174-5e99d96a-8ce8-4b0a-b6e8-849214126a2f.png">
cinic这里调用的是cifar_noniid_s()吗？

################

然后CINIC10数据集划分这部分的代码，
<img width="1242" alt="image" src="https://user-images.githubusercontent.com/90205287/237024544-97774745-d4d2-44e3-87ea-39236341d349.png">
这部分是咋回事呀，为什么有cinic10_cifar和cinic10_imagenet两个呀。。。
这部分对应论文的哪部分实验呀。。

再次分隔线————————————————————————————————————————————————————

最后再弱弱的问一句，为啥在实验部分没考虑CIFAR100数据集的对比实验呢。。。 


——————————————————————————————————————————————————————————

问的太多有点不好意思了。。 也不知道怎么报答老哥的善意  QAQ

**Comment by Bhongcheng** [comment]
这是我重新跑的结果。。。

emnist:
    FedAvg-FT:  84.29   86.35
    FedPAC:     88.38   89.94

fmnist
    FedAvg-FT:  90.21   91.72
    FedPAC:     91.73   93.07

cifar10:
    FedAvg-FT:  76.48   80.33
    FedPAC:     81.02   84.92

cinic10:
    FedAvg-FT:  57.20   57.38
    FedPAC:     63.50   66.05

**Comment by JianXu95** [comment]
cinic数据集是由cifar和imagenet子集组合而成，我实际实验中应该是分开来读取了，就是各一半的client分别读俩子集，这样子feature分布也稍稍有点不一样，然后事实证明也work，所以cinic这个数据集提供了多种数据加载方法. code数据加载那部分写的应该也没问题。你先试试cinic_sep吧。没跑cifar100是因为类别太多了普通cnn效果太差，复杂一点的模型都有bn层就暂时搁置了（以及并没有大显存gpu跑大一点模型= =，科研条件艰苦😥）发自我的 iPhone在 2023年5月9日，06:19，鲍鸿丞 ***@***.***> 写道：﻿
这是我重新跑的结果。。。
emnist:
FedAvg-FT:  84.29   86.35
FedPAC:     88.38   89.94
fmnist
FedAvg-FT:  90.21   91.72
FedPAC:     91.73   93.07
cifar10:
FedAvg-FT:  76.48   80.33
FedPAC:     81.02   84.92
cinic10:
FedAvg-FT:  57.20   57.38
FedPAC:     63.50   66.05

**Comment by Bhongcheng** [comment]
老哥，我再弱弱的问下。。我下载的CINIC-10数据集解压后文件目录是这样的
<img width="354" alt="image" src="https://github.com/JianXu95/FedPAC/assets/90205287/f349f193-329c-4a36-86bc-ea33f421bf3e">
而老哥你代码里的文件目录是这样的
<img width="512" alt="image" src="https://github.com/JianXu95/FedPAC/assets/90205287/d5e5a40f-3d82-4027-9684-704a30b67c9b">
<img width="534" alt="image" src="https://github.com/JianXu95/FedPAC/assets/90205287/f4628147-d08f-4a1d-a32b-0ce70ea6def9">

这个是自己手动划分的吗？ 还是说有可以直接下载对应版本的CINIC-10数据集呢？

谢谢老哥

**Comment by JianXu95** [comment]
啊你不提我都忘了这个细节，是我事先自己写代码划分生成的，这部分代码不复杂网上有类似现成的，好像不在工程文件里，我回头找找= =发自我的 iPhone在 2023年5月10日，00:11，鲍鸿丞 ***@***.***> 写道：﻿
老哥，我再弱弱的问下。。我下载的CINIC-10数据集解压后文件目录是这样的

而老哥你代码里的文件目录是这样的


这个是自己手动划分的吗？ 还是说有可以直接下载对应版本的CINIC-10数据集呢？
谢谢老哥

**Comment by Pengxin-Guo** [comment]
请问你最后CINIC-10数据集是如何处理的呢？我也是直接从网上下载的CINIC-10，跑的结果很差。请问是需要分别下载CIFAR-10和ImageNet自己处理吗？

**Comment by Bhongcheng** [comment]
不知道呀

**Comment by Pengxin-Guo** [comment]
@JianXu95 请问老哥可以将CINIC-10-CIFAR和CINIC-10-ImageNet里面的数据也开源了吗？非常感谢！

**Comment by JianXu95** [comment]
The original cinic dataset should be the same, from which I used some codes to split it into two subsets of `cifar` and `imagenet`, I attached the codes below, you can copy it for a try. The codes are partially borrowed from the internet.

    symlink = False  # If this is false the files are copied instead
    combine_train_valid = False  # If this is true, the train and valid sets are ALSO combined
    cinic_directory = "data/CINIC-10"
    imagenet_directory = "data/CINIC-10-ImageNet"
    cifar_directory = "data/CINIC-10-CIFAR"
    classes = ["airplane", "automobile", "bird", "cat", "deer", "dog", "frog", "horse", "ship", "truck"]
    sets = ['train', 'valid', 'test']
    directory = imagenet_directory  # or cifar_directory
    if not os.path.exists(directory):
        os.makedirs(directory)
    if not os.path.exists(directory + '/train'):
        os.makedirs(directory + '/train')
    if not os.path.exists(directory + '/test'):
        os.makedirs(directory + '/test')

    for c in classes:
        if not os.path.exists('{}/train/{}'.format(directory, c)):
            os.makedirs('{}/train/{}'.format(directory, c))
        if not os.path.exists('{}/test/{}'.format(directory, c)):
            os.makedirs('{}/test/{}'.format(directory, c))
        if not combine_train_valid:
            if not os.path.exists('{}/valid/{}'.format(directory, c)):
                os.makedirs('{}/valid/{}'.format(directory, c))

    for s in sets:
        for c in classes:
            source_directory = '{}/{}/{}'.format(cinic_directory, s, c)
            filenames = glob.glob('{}/*.png'.format(source_directory))
            for fn in filenames:
                dest_fn = fn.split('/')[-1]
                if (s == 'train' or s == 'valid') and combine_train_valid and 'cifar' not in fn.split('/')[-1]:
                    dest_fn = '{}/train/{}/{}'.format(directory, c, dest_fn)
                    if symlink:
                        if not os.path.islink(dest_fn):
                            os.symlink(fn, dest_fn)
                    else:
                        copyfile(fn, dest_fn)

                elif (s == 'train') and 'cifar' not in fn.split('/')[-1]:
                    dest_fn = '{}/train/{}/{}'.format(directory, c, dest_fn)
                    if symlink:
                        if not os.path.islink(dest_fn):
                            os.symlink(fn, dest_fn)
                    else:
                        copyfile(fn, dest_fn)

                elif (s == 'valid') and 'cifar' not in fn.split('/')[-1]:
                    dest_fn = '{}/valid/{}/{}'.format(directory, c, dest_fn)
                    if symlink:
                        if not os.path.islink(dest_fn):
                            os.symlink(fn, dest_fn)
                    else:
                        copyfile(fn, dest_fn)

                elif s == 'test' and 'cifar' not in fn.split('/')[-1]:
                    dest_fn = '{}/test/{}/{}'.format(directory, c, dest_fn)
                    if symlink:
                        if not os.path.islink(dest_fn):
                            os.symlink(fn, dest_fn)
                    else:
                        copyfile(fn, dest_fn)

**Comment by Pengxin-Guo** [comment]
Thank you very much. I have reproduced the results in the paper according to your prompt. Thanks again.

**Comment by SanaAwan5** [comment]
I am trying to run the code with different datasets and methods. Also keeping the settings default. But it seem that model start to overfit from the second round and performance doesn't improve. Can you please see if any changes has been made in code recently

**Comment by SanaAwan5** [comment]
<img width="557" alt="Screen Shot 2023-11-30 at 3 38 37 PM" src="https://github.com/JianXu95/FedPAC/assets/47128808/0c866c78-f582-47ea-b675-f330ef719d2c">

**Comment by gjhghjh6** [comment]
你好，代码压缩包能发一份吗，github上没了
